service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/Query.java

```diff
@@ -335,13 +335,192 @@ public QueryResponse sendQueryRequest(IotHubConnectionString iotHubConnectionStr
+
+        this.proxy = proxy;
+
+        byte[] payload;
+        Map<String, String> queryHeaders = new HashMap<>();
+
+        if (this.requestContinuationToken != null)
+        {
+            queryHeaders.put(CONTINUATION_TOKEN_KEY, requestContinuationToken);
+        }
+        queryHeaders.put(PAGE_SIZE_KEY, String.valueOf(pageSize));
+
+        DeviceOperations.setHeaders(queryHeaders);
+
+        if (isSqlQuery)
+        {
+            QueryRequestParser requestParser = new QueryRequestParser(this.query);
+            payload = requestParser.toJson().getBytes();
+        }
+        else
+        {
+            payload = new byte[0];
+        }
+
+        HttpResponse httpResponse = DeviceOperations.request(azureSasCredential.getSignature(), url, method, payload, null, httpConnectTimeout, httpReadTimeout, proxy);
+
+        this.responseContinuationToken = null;
+        Map<String, String> headers = httpResponse.getHeaderFields();
+        for (Map.Entry<String, String> header : headers.entrySet())
+        {
+            switch (header.getKey())
+            {
+                case CONTINUATION_TOKEN_KEY:
+                    this.responseContinuationToken = header.getValue();
+                    break;
+                case ITEM_TYPE_KEY:
+                    this.responseQueryType = QueryType.fromString(header.getValue());
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        if (this.responseQueryType == null || this.responseQueryType == QueryType.UNKNOWN)
+        {
+            throw new IOException("Query response type is not defined by IotHub");
+        }
+
+        if (this.requestQueryType != this.responseQueryType)
+        {
+            throw new IOException("Query response does not match query request");
+        }
+
+        this.queryResponse = new QueryResponse(new String(httpResponse.getBody()));
+        return this.queryResponse;
+    }
+
+    /**
+     * Sends request for the query to the IotHub.
+     *
+     * @param credential The authorization token provider.
+     * @param url URL to Query on.
+     * @param method HTTP Method for the requesting a query.
+     * @param httpConnectTimeout the http connect timeout to use for this request.
+     * @param httpReadTimeout the http read timeout to use for this request.
+     * @param proxy the proxy to use, or null if no proxy should be used.
+     * @return QueryResponse object which holds the response Iterator.
+     * @throws IOException If any of the input parameters are not valid.
+     * @throws IotHubException If HTTP response other then status ok is received.
+     */
+    public QueryResponse sendQueryRequest(
+            TokenCredential credential,
+            URL url,
+            HttpMethod method,
+            int httpConnectTimeout,
+            int httpReadTimeout,
+            Proxy proxy)
+            throws IOException, IotHubException
```

Comment:
Why IOException?

Reply:
This method sends an HTTP request which may throw an IOException